plugins/toolbox-search/test/tests.mocha.js

```diff
@@ -57,6 +66,119 @@ suite('BlockSearcher', () => {
     assert.sameMembers(ransomNoteMatches, [listCreateWithBlock]);
   });
 
+  test('requires the final trigram when matching longer queries', () => {
+    const searcher = new BlockSearcher();
+    const mathConstrainBlock = {
+      kind: 'block',
+      type: 'math_constrain',
+    };
+    searcher.indexBlocks([mathConstrainBlock]);
+
+    const matches = searcher.blockTypesMatching('conso');
+
+    assert.notInclude(
+      matches,
+      mathConstrainBlock,
+      'query missing trailing trigram should not match',
+    );
+  });
+
+  test('normalizes underscores in block types to spaces', () => {
+    if (!Blockly.Blocks['searcher_underscore_block']) {
+      Blockly.defineBlocksWithJsonArray([
+        {
+          type: 'searcher_underscore_block',
+          message0: 'custom block with underscore',
+        },
+      ]);
+    }
+
+    const searcher = new BlockSearcher();
+    const blockInfo = {
+      kind: 'block',
+      type: 'searcher_underscore_block',
+    };
+    searcher.indexBlocks([blockInfo]);
+
+    assert.sameMembers(
+      searcher.blockTypesMatching('custom block with underscore'),
+      [blockInfo],
+    );
+    assert.isEmpty(searcher.blockTypesMatching('custom_block_with_underscore'));
+  });
+
+  test('longer queries disambiguate similar blocks', () => {
+    if (!Blockly.Blocks['searcher_charlie']) {
+      Blockly.defineBlocksWithJsonArray([
+        {
+          type: 'searcher_charlie',
+          message0: 'alpha bravo charlie',
+        },
+        {
+          type: 'searcher_delta',
+          message0: 'alpha bravo delta',
+        },
+      ]);
+    }
+
+    const searcher = new BlockSearcher();
+    const blockA = {kind: 'block', type: 'searcher_charlie'};
+    const blockB = {kind: 'block', type: 'searcher_delta'};
+
+    searcher.indexBlocks([blockA, blockB]);
+
+    const broadQueryMatches = searcher.blockTypesMatching('alpha bravo');
+    assert.sameMembers(broadQueryMatches, [blockA, blockB]);
+
+    const specificQueryMatches =
+      searcher.blockTypesMatching('alpha bravo charlie');
+    assert.sameMembers(specificQueryMatches, [blockA]);
+  });
+
+  test('indexes dropdown alt text options', () => {
+    if (!Blockly.Blocks['searcher_dropdown_alt']) {
+      Blockly.defineBlocksWithJsonArray([
+        {
+          type: 'searcher_dropdown_alt',
+          message0: 'weather %1',
+          args0: [
+            {
+              type: 'field_dropdown',
+              name: 'WEATHER',
+              options: [
+                [
+                  {
+                    src: 'data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEA',
```

Comment:
This doesn't appear to actually be a valid image; if it doesn't otherwise cause problems, can you set `src` to `''` here and below to avoid confusion as to what this is and whether the particular value is relevant to the test case?